rmw_connextdds_common/src/common/rmw_context.cpp

```diff
@@ -109,35 +109,44 @@ rmw_connextdds_initialize_participant_qos(
 
 rmw_ret_t
 rmw_context_impl_t::initialize_node(
-  const char * const node_name,
-  const char * const node_namespace,
-  const bool localhost_only)
+  const rmw_discovery_options_t * const discovery_options)
 {
-  UNUSED_ARG(node_name);
-  UNUSED_ARG(node_namespace);
-
-  RMW_CONNEXT_LOG_DEBUG_A(
-    "initializing new node: total=%lu, localhost=%d",
-    this->node_count, localhost_only)
-
   if (0u != this->node_count) {
-    if ((this->localhost_only && !localhost_only) ||
-      (!this->localhost_only && localhost_only))
-    {
-      RMW_CONNEXT_LOG_ERROR_A_SET(
-        "incompatible node for context:"
-        "ctx.localhost_only=%d, node.localhost_only=%d",
-        this->localhost_only, localhost_only)
+    bool params_equal = false;
+    if (rmw_discovery_options_equal(this->discovery_options, discovery_options, &params_equal) != RMW_RET_OK) {
+      RMW_CONNEXT_LOG_ERROR_SET("invalid discovery params argument");
+      return RMW_RET_INVALID_ARGUMENT;
+    }
+
+    if (!params_equal) {
+      RMW_CONNEXT_LOG_ERROR_SET(
+        "node is being initialized with incompatible discovery parameters");
       return RMW_RET_ERROR;
     }
 
     this->node_count += 1;
     RMW_CONNEXT_LOG_DEBUG_A(
-      "initialized new node: total=%lu", this->node_count)
+      "initialized new node: total=%lu", this->node_count);
     return RMW_RET_OK;
   }
 
-  rmw_ret_t rc = this->initialize_participant(localhost_only);
+  if (!this->discovery_options && discovery_options) {
+    this->discovery_options = (rmw_discovery_options_t*)
+      this->base->options.allocator.allocate(
+      sizeof(rmw_discovery_options_t),
+      this->base->options.allocator.state);
+
+    const auto rc = rmw_discovery_options_copy(
+      discovery_options,
+      this->base->options.allocator,
+      this->discovery_options);
+    if (RMW_RET_OK != rc) {
+      RMW_CONNEXT_LOG_ERROR("failed to copy discovery parameters");
```

Comment:
This should check the error message and add it to the log message. Or better yet, add to the error message and return the return code. Unconditionally logging here doesn't seem like exactly the right thing to do.